src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+            SOL_INT_CHECK(r, < 0, r);
+
+            if (service->name)
+                free(service->name);
+            service->name = strdup(name);
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+            SOL_NULL_CHECK(service->name, -ENOMEM);
+        } else if (strcmp(str, "State") == 0) {
+            char *state;
+            static const struct sol_str_table table[] = {
+                SOL_STR_TABLE_ITEM("online",
+                    SOL_CONNMAN_SERVICE_STATE_ONLINE),
+                SOL_STR_TABLE_ITEM("ready",
+                    SOL_CONNMAN_SERVICE_STATE_READY),
+                SOL_STR_TABLE_ITEM("association",
+                    SOL_CONNMAN_SERVICE_STATE_ASSOCIATION),
+                SOL_STR_TABLE_ITEM("configuration",
+                    SOL_CONNMAN_SERVICE_STATE_CONFIGURATION),
+                SOL_STR_TABLE_ITEM("disconnect",
+                    SOL_CONNMAN_SERVICE_STATE_DISCONNECT),
+                SOL_STR_TABLE_ITEM("idle",
+                    SOL_CONNMAN_SERVICE_STATE_IDLE),
+                SOL_STR_TABLE_ITEM("failure",
+                    SOL_CONNMAN_SERVICE_STATE_FAILURE),
+                { }
+            };
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &state);
+            SOL_INT_CHECK(r, < 0, r);
+
+            if (state)
+                service->state = sol_str_table_lookup_fallback(table,
+                    sol_str_slice_from_str(state),
+                    SOL_CONNMAN_SERVICE_STATE_UNKNOWN);
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (strcmp(str, "Strength") == 0) {
+            uint8_t strength = 0;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "y");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, SD_BUS_TYPE_BYTE, &strength);
+            SOL_INT_CHECK(r, < 0, r);
+
+            service->strength = strength;
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else if (strcmp(str, "Type") == 0) {
+            char *type;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &type);
+            SOL_INT_CHECK(r, < 0, r);
+
+            if (service->type)
+                free(service->type);
+            service->type = strdup(type);
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+            SOL_NULL_CHECK(service->type, -ENOMEM);
+        } else if (strcmp(str, "IPv4") == 0) {
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "a{sv}");
+            SOL_INT_CHECK(r, < 0, r);
+
+            get_service_ipv4(m, service);
```

Comment:
I'm wondering if we should store the gateway address as well. What about netmask?

Do you know if these would be used for an application?